sdk/src/main/scala/com/horizen/certificatesubmitter/CertificateSubmitter.scala

```diff
@@ -0,0 +1,98 @@
+package com.horizen.certificatesubmitter
+
+
+import akka.actor.{Actor, ActorRef, ActorSystem, Props}
+import akka.pattern.ask
+import akka.util.Timeout
+import com.horizen.SidechainNodeViewHolder.ReceivableMessages.GetDataFromCurrentSidechainNodeView
+import com.horizen._
+import com.horizen.block.SidechainBlock
+import com.horizen.box.WithdrawalRequestBox
+import com.horizen.mainchain.{CertificateRequest, CertificateRequestCreator}
+import com.horizen.mainchain.api.RpcMainchainApi
+import com.horizen.node.SidechainNodeView
+import com.horizen.params.NetworkParams
+import com.horizen.utils.WithdrawalEpochUtils
+import scorex.core.NodeViewHolder.CurrentView
+import scorex.core.network.NodeViewSynchronizer.ReceivableMessages.SemanticallySuccessfulModifier
+import scorex.util.ScorexLogging
+
+import scala.collection.JavaConverters._
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success, Try}
+
+class CertificateSubmitter
+  (settings: SidechainSettings,
+   sidechainNodeViewHolderRef: ActorRef,
+   params: NetworkParams)
+  (implicit ec: ExecutionContext)
+  extends Actor
+  with ScorexLogging
+{
+
+  val timeoutDuration: FiniteDuration = settings.scorexSettings.restApi.timeout
+  implicit val timeout: Timeout = Timeout(timeoutDuration)
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[SemanticallySuccessfulModifier[SidechainBlock]])
+  }
+
+  lazy val mainchainApi = new RpcMainchainApi(settings)
+
+  protected def viewAsync(): Future[SidechainNodeView] = {
+    def f(v: SidechainNodeView) = v
+
+    (sidechainNodeViewHolderRef ? GetDataFromCurrentSidechainNodeView(f))
+      .mapTo[SidechainNodeView]
+  }
+
+  protected def trySubmitCertificate: Unit = {
+    viewAsync().onComplete {
+      case Success(sidechainNodeView) => {
+        val withdrawalEpochInfo = sidechainNodeView.getNodeState.getWithdrawalEpochInfo
+        if (withdrawalEpochInfo.epoch > 0 &&
+            WithdrawalEpochUtils.canSubmitCertificate(withdrawalEpochInfo, params)) {
+          val withdrawalRequests = sidechainNodeView.getNodeState.getWithdrawalRequests(withdrawalEpochInfo.epoch - 1)
+          if (withdrawalRequests.isPresent) {
```

Comment:
I think no. We have three cases:
1. Certificate was sent and returned back from mainchain - withdrawalRequests contains None
2. Certificate was not sent, but list of withdrawal requests for previous epoch is emty - withdrawalRequests contains Some(empty sequence)
3. Certificate was not sent and list of withdrawal requests for previous epoch is not emty - withdrawalRequests contains Some(non-empty sequence)

Reply:
Ok, I get it.
Main logical problem here, that `sidechainNodeView.getNodeState.getWithdrawalRequests` actually means: return me requests for that epoch if the cert is not present yet.
`sidechainNodeView.getNodeState.getUnprocessedWithdrawalRequests` is much clear.
`getWithdrawalRequests` must return the seq in any case.

As for me, just retrieving the boolean flag `state.containsCertificateForEpoch(epoch)` will simplify a lot, instead of having `getWithdrawalRequests` and `getUnprocessedWithdrawalRequests` or just one with mixed logic.